### PR TITLE
[Datasets] Fix zip stage to preserve order when executing the other side

### PR DESCRIPTION
**Pull request by c21**
## Why are these changes needed?

Found bug that `ZipStage` not preserve order when executing the right side. Though previously we have check if plan has zip or sort stage, that only enforces order from left side, but not enforces order from right side.

See the added unit test for a reproduce example. User encounter the issue when enabling streaming execution on their workload.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/execution/legacy_compat.py [hunk 1]
```diff
@@ -109,6 +116,7 @@ def _get_execution_dag(
     executor: Executor,
     plan: ExecutionPlan,
     allow_clear_input_blocks: bool,
+    preserve_order: bool = False,
```
Comment (jianoaix) [inline python/ray/data/_internal/execution/legacy_compat.py:119]: Shall we make it required as other params and as upper call funcs?

**Review by jianoaix — APPROVED** [review]

File: python/ray/data/_internal/execution/legacy_compat.py (hunk 1 above)
Comment (c21) [reply python/ray/data/_internal/execution/legacy_compat.py:119]: yeah we can make it required. sorry what do you mean by "as upper call funcs"?

Comment (jianoaix) [reply python/ray/data/_internal/execution/legacy_compat.py:119]: Oh, the callers of this function, like `execute_to_legacy_block_list`.

Comment (c21) [reply python/ray/data/_internal/execution/legacy_compat.py]: Got it, updated.

**Pull request merged**